cpp/daal/src/algorithms/dtrees/gbt/gbt_predict_dense_default_impl.i

```diff
@@ -155,24 +156,36 @@ struct TileDimensions
     size_t nCols         = 0;
     size_t nRowsInBlock  = 0;
     size_t nTreesInBlock = 0;
+    size_t nLargeBlocks  = 0;
     size_t nDataBlocks   = 0;
     size_t nTreeBlocks   = 0;
 
-    TileDimensions(const NumericTable & data, size_t nTrees)
+    // vectorBlockSize = vectorBlockSizeFactor * vectorBlockSizeStep
+    size_t vectorBlockSizeFactor                     = 0;
+    static constexpr size_t maxVectorBlockSizeFactor = 16;
+    static constexpr size_t minVectorBlockSizeFactor = 2;
+    static constexpr size_t vectorBlockSizeStep      = 16;
+    // optimalBlockSizeFactor is selected from benchmarking
+    static constexpr size_t optimalBlockSizeFactor = 3;
+
+    TileDimensions(const NumericTable & data, size_t nTrees, size_t nNodes)
         : nTreesTotal(nTrees), nRowsTotal(data.getNumberOfRows()), nCols(data.getNumberOfColumns())
     {
-        nRowsInBlock = nRowsTotal;
-
-        if (nRowsTotal > 2 * VECTOR_BLOCK_SIZE)
-        {
-            nRowsInBlock = 2 * VECTOR_BLOCK_SIZE;
-
-            if (daal::threader_get_threads_number() > nRowsTotal / nRowsInBlock)
-            {
-                nRowsInBlock = VECTOR_BLOCK_SIZE;
-            }
-        }
-        nDataBlocks = nRowsTotal / nRowsInBlock;
+        // Use smaller vectorBlockSize if trees fit to L2
+        // Each node contain 3 values
+        size_t nodesSize                = (sizeof(ModelFPType) + sizeof(FeatureIndexType) + sizeof(int)) * nNodes;
+        constexpr float add_hoc_L2_size = 0.85 * 2 * 1024 * 1024;
```

Comment:
Please use daal::services::internal::getL2CacheSize() to define the size of L2 cache.
There are also other functions you might find useful in "service_envirinment.h"

Reply:
done